// @pulumi/hcloud/server.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
const pulumi = require("@pulumi/pulumi");
class Server extends pulumi.CustomResource {
    /**
     * Get an existing Server resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     */
    static get(name, id, state, opts) {
        return new Server(name, state, Object.assign({}, opts, { id: id }));
    }
    constructor(name, argsOrState, opts) {
        let inputs = {};
        if (opts && opts.id) {
            const state = argsOrState;
            inputs["backupWindow"] = state ? state.backupWindow : undefined;
            inputs["backups"] = state ? state.backups : undefined;
            inputs["datacenter"] = state ? state.datacenter : undefined;
            inputs["image"] = state ? state.image : undefined;
            inputs["ipv4Address"] = state ? state.ipv4Address : undefined;
            inputs["ipv6Address"] = state ? state.ipv6Address : undefined;
            inputs["ipv6Network"] = state ? state.ipv6Network : undefined;
            inputs["iso"] = state ? state.iso : undefined;
            inputs["keepDisk"] = state ? state.keepDisk : undefined;
            inputs["labels"] = state ? state.labels : undefined;
            inputs["location"] = state ? state.location : undefined;
            inputs["name"] = state ? state.name : undefined;
            inputs["rescue"] = state ? state.rescue : undefined;
            inputs["serverType"] = state ? state.serverType : undefined;
            inputs["sshKeys"] = state ? state.sshKeys : undefined;
            inputs["status"] = state ? state.status : undefined;
            inputs["userData"] = state ? state.userData : undefined;
        }
        else {
            const args = argsOrState;
            if (!args || args.image === undefined) {
                throw new Error("Missing required property 'image'");
            }
            if (!args || args.serverType === undefined) {
                throw new Error("Missing required property 'serverType'");
            }
            inputs["backups"] = args ? args.backups : undefined;
            inputs["datacenter"] = args ? args.datacenter : undefined;
            inputs["image"] = args ? args.image : undefined;
            inputs["iso"] = args ? args.iso : undefined;
            inputs["keepDisk"] = args ? args.keepDisk : undefined;
            inputs["labels"] = args ? args.labels : undefined;
            inputs["location"] = args ? args.location : undefined;
            inputs["name"] = args ? args.name : undefined;
            inputs["rescue"] = args ? args.rescue : undefined;
            inputs["serverType"] = args ? args.serverType : undefined;
            inputs["sshKeys"] = args ? args.sshKeys : undefined;
            inputs["userData"] = args ? args.userData : undefined;
            inputs["backupWindow"] = undefined /*out*/;
            inputs["ipv4Address"] = undefined /*out*/;
            inputs["ipv6Address"] = undefined /*out*/;
            inputs["ipv6Network"] = undefined /*out*/;
            inputs["status"] = undefined /*out*/;
        }
        super("hcloud:index/server:Server", name, inputs, opts);
    }
}
exports.Server = Server;
